pilot/pkg/trustbundle/trustbundle_test.go

```diff
@@ -0,0 +1,135 @@
+	}
+}
+
+func TestVerifyTrustAnchor(t *testing.T) {
+	testCases := []struct {
+		errExp bool
+		cert   string
+	}{
+		{
+			cert:   testCerts["RootCACert"],
+			errExp: false,
+		},
+		{
+			cert:   testCerts["MalformedCert"],
+			errExp: true,
+		},
+		{
+			cert:   testCerts["NonCaCert"],
+			errExp: true,
+		},
+		{
+			cert:   testCerts["IntermediateCACert"],
+			errExp: false,
+		},
+	}
+	for i, tc := range testCases {
+		err := NewTrustBundle(func() {}).verifyTrustAnchor(tc.cert)
+		if tc.errExp && err == nil {
+			t.Errorf("test case %v failed. Expected Error but got none", i)
+		} else if !tc.errExp && err != nil {
+			t.Errorf("test case %v failed. Expected no error but got %v", i, err)
+		}
+	}
+}
+
+func TestUpdateTrustAnchor(t *testing.T) {
+	var cbCounter int = 0
+	tb := NewTrustBundle(func() { cbCounter++ })
+	var trustedCerts []string
+
+	// Add First Cert update
+	tb.UpdateTrustAnchor(&TrustAnchorUpdate{TrustAnchorConfig{Source: SourceMeshConfig, Certs: []string{testCerts["RootCACert"]}}})
+	trustedCerts = tb.GetTrustBundle()
+	if !checkSameCerts(trustedCerts, []string{testCerts["RootCACert"]}) || cbCounter != 1 {
+		t.Errorf("Basic trustbundle update test failed. Callback value is %v", cbCounter)
+	}
+
+	// Add Second Cert update
+	// ensure intermediate CA certs accepted, it replaces the first completely, and lib dedupes duplicate cert
+	tb.UpdateTrustAnchor(&TrustAnchorUpdate{TrustAnchorConfig{
+		Source: SourceMeshConfig,
+		Certs:  []string{testCerts["IntermediateCACert"], testCerts["IntermediateCACert"]},
+	}})
+	trustedCerts = tb.GetTrustBundle()
+	if !checkSameCerts(trustedCerts, []string{testCerts["IntermediateCACert"]}) || cbCounter != 2 {
+		t.Errorf("trustbundle intermediate cert update test failed. Callback value is %v", cbCounter)
+	}
+
+	// Try adding one more cert to a different source
+	// Ensure both certs are not present
+	tb.UpdateTrustAnchor(&TrustAnchorUpdate{TrustAnchorConfig{Source: SourceIstioCA, Certs: []string{testCerts["RootCACert"]}}})
+	trustedCerts = tb.GetTrustBundle()
+	sort.Strings(trustedCerts)
+	result := []string{testCerts["IntermediateCACert"], testCerts["RootCACert"]}
+	sort.Strings(result)
+	if !checkSameCerts(trustedCerts, result) || cbCounter != 3 {
+		t.Errorf("multicert update failed. Callback value is %v", cbCounter)
+	}
+
+	// Try added same cert again. Ensure cb doesn't increment
+	tb.UpdateTrustAnchor(&TrustAnchorUpdate{TrustAnchorConfig{Source: SourceIstioCA, Certs: []string{testCerts["RootCACert"]}}})
+	trustedCerts = tb.GetTrustBundle()
+	sort.Strings(trustedCerts)
+	if !checkSameCerts(trustedCerts, result) || cbCounter != 3 {
+		t.Errorf("duplicate multicert update failed. Callback value is %v", cbCounter)
+	}
+
+	// Try added one good cert, one bogus Cert
+	// Verify Update should not go through and no change to cb
+	tb.UpdateTrustAnchor(&TrustAnchorUpdate{TrustAnchorConfig{Source: SourceIstioCA, Certs: []string{testCerts["Malformed"]}}})
```

Comment:
should we check the error and make sure it gives one? And possibly make sure all other cases do not?